### ISSUE TITLE
'Strange' results when decoding

### ISSUE THREAD

**Issue by chris-aeviator** [hannamw/EAP-IG#4]
When decoding the logits inside evalutate_graph, I get results that seem way off the original functionality even though the evaluation metric suggests a similar result to the vanilla model.

``
Original performance was 0.8143222332000732; the circuit's performance is 0.7321224212646484
``

eval_graph:
```
 first was for January beginning 200076 to the year 1353
  2%|███▊                                                                                                                                                            | 2/84 [00:00<00:28,  2.90it/s]

 first was for January beginning 200045 to the year 1549
  4%|█████▋                                                                                                                                                          | 3/84 [00:01<00:27,  2.98it/s]

 first information for about moment before50 to the year 1370
  5%|███████▌                                                                                                                                                        | 4/84 [00:01<00:26,  3.07it/s]
 first of about January beginning 200076 to the year 1750
```

vanilla:
```

>>> model.generate('The contract lasted from the year 1352 to the year')
'The contract lasted from the year 1352 to the year 2372. The Yeonsu included arguments for'
```

Am I misunderstanding something here?

**Comment by hannamw** [comment]
Hi! I'm not sure exactly what experiment you've run here, but generally, finding a circuit that achieves high loss / metric recovery doesn't mean that the circuit characterizes an entire generation well. That is, the circuit aims to explain only the value of the metric computed on the next-token distribution given the task input. So if you're performing generation, which involves repeatedly computing the next-token distribution, sampling from it, and then recomputing the next-token distribution on a new input, the circuit will only explain the first step of the generation. 

Imagine the greater-than circuit explains your model's (metric's) behavior on "The war lasted from the year 1741 to the year 17". Let's say that you then sample a next token "52". The circuit won't explain your model's behavior on "The war lasted from the year 1741 to the year 1752", where it's probably not performing greater-than / predicting a number.

Also, depending on the metric you use, the metric might not take into account the entire probability distribution; the metric could look good, while the rest of the distribution could be different. Prob diff on greater-than considers only a small part of the next-token distribution, while other metrics (KL divergence, for example), look at the whole distribution. That said if your prob diff is high (e.g. 0.81, as in the paper) and you recover e.g. 0.9 of it, your circuit will explain the majority of the distribution.

If you still think there's an issue, would you mind providing a minimal example to reproduce this - something like the model / circuit you're using, as well as the code and inputs you're using to produce these unexpected outputs?
